saw-core/src/Verifier/SAW/FiniteValue.hs

```diff
@@ -336,9 +348,21 @@ scFirstOrderValue sc fv =
     FOVVec t vs -> do t' <- scFirstOrderType sc t
                       vs' <- traverse (scFirstOrderValue sc) vs
                       scVector sc t' vs'
-    FOVArray t1 t2 -> do t1' <- scFirstOrderType sc t1
-                         t2' <- scFirstOrderType sc t2
-                         scArrayType sc t1' t2'
+    FOVArray tk d vs -> do
+        -- first get the key and value types
+        tk' <- scFirstOrderType sc tk
+        tv' <- scFirstOrderType sc $ firstOrderTypeOf d
+        -- convert the default value and make an array
+        d' <- scFirstOrderValue sc d
+        arr0 <- scArrayConstant sc tk' tv' d'
+        -- now add each update to the array (monadic fold)
+        let visit k v marr = do
+              arr <- marr
+              k' <- scFirstOrderValue sc k
+              v' <- scFirstOrderValue sc v
+              scArrayUpdate sc tk' tv' arr k' v'
+        arr <- Map.foldrWithKey visit (pure arr0) vs
```

Comment:
If you use [`ifoldrM`](https://hackage.haskell.org/package/indexed-traversable-0.1.4/docs/Data-Foldable-WithIndex.html#v:ifoldrM) (from the `indexed-traversable` package, which `saw-script` already depends on), then you can avoid needing the extra `pure` here.

Reply:
done